test/test_entrypoint.py

```diff
@@ -0,0 +1,297 @@
+                patch_update_status.start()
+
+                ep.entrypoint()
+                # re-emit should pick the deferred config-changed
+                self.assertEqual(CharmConfigDeferrer.count_called, 2)
+                self.assertEqual(CharmConfigDeferrer.defer, False)
+
+    @patch('juju.entrypoint.setup_hooks')
+    def test_all_base_events_handled(self, _setup_hooks):
+        # note: class variables are modified because Charm instances
+        # are not preserved across simulated hook executions
+        class Charm(juju.charm.CharmBase):
+            count_called = 0
+
+            events_to_observe = []
+            events_types_observed = set()
+
+            on = juju.charm.CharmEventsBase()
+
+            def __init__(self, framework):
+                super().__init__(framework)
+                for event_name in self.events_to_observe:
+                    self.framework.observe(getattr(self.on, event_name), self)
+
+            def _handler_core(self, event):
+                logger.debug(f'In {inspect.stack()[0].function},'
+                             f' event {event}')
+                type(self).count_called += 1
+                type(self).events_types_observed.add(type(event))
+
+        # infer supported "base" events from the base charm class
+        base_events = {event_name: instance for event_name, instance
+                       in juju.charm.CharmEventsBase.__dict__.items()
+                       if type(instance) == juju.framework.Event}
+
+        event_names = base_events.keys()
+        event_types = {e.event_type for e in base_events.values()}
+
+        logger.debug(f'Expected event types {event_types}')
+
+        # make sure all of the base events are observed
+        Charm.events_to_observe = event_names
+
+        # set up handlers for all base events reusing one implementation
+        for event_name in event_names:
+            setattr(Charm, f'on_{event_name}', Charm._handler_core)
+
+        # simulate hook executions for every event
+        for event_name in event_names:
+            with patch('sys.argv', [event_name]):
+                with patch('juju.entrypoint.Charm', Charm):
+                    ep.entrypoint()
+
+        logger.debug(f'Events types observed {Charm.events_types_observed}')
+
+        self.assertEqual(len(event_names), Charm.count_called)
+        self.assertEqual(Charm.events_types_observed, event_types)
+
+    def test_event_not_implemented(self):
+        """Make sure Juju events are not silently skipped.
+        """
+        class Charm(juju.charm.CharmBase):
+            on = juju.charm.CharmEventsBase()
+
+        with patch('juju.entrypoint.Charm', Charm):
+            with patch('sys.argv', ['not-implemented-event']):
+                self.assertRaises(NotImplementedError, ep.entrypoint)
+
+    @patch('juju.entrypoint.load_metadata')
+    @patch('os.path.exists')
+    @patch('os.unlink')
+    @patch('os.symlink')
```

Comment:
The rest of this PR looks good. But the amount of patching and fiddling in the testing still makes me uncomfortable. Let's have a meeting tomorrow and try to sort this out together so we can get this PR merged.

Reply:
I agree, thanks for tracking!